### DISCUSSION TITLE
FGCOM with and without MSFS

### DISCUSSION THREAD

**Discussion by didierma34** [hbeni/fgcom-mumble#159] [Q&A] [answered]
Hello,

Is it possible to use radio FGCOM Java without launching MSFS for an ATC ?

Thank's
Didier

**Comment by hbeni** [comment, marked as answer]
Hi, yes it is.
The FGCom-mumble RadioGUI can be used exactly for that. You can setup your location and radio stack there.
Only pilots using FGCom-mumble can participate, however, and they need the Simconnect bridging feature of the RadioGUI for that.

**Comment by didierma34** [comment]
Ok, Thank's a lot.

Tomorow I'll receive my Linux so I will try to install Mumble on it.
I'll have an Ubuntu. Is-it eays to install Mumble on it ?

Thank's for your help :) :) :)

**Reply by hbeni** [reply]
It is sort of easy. Afaik there is not yet a package for it, so you need to self compile.
The mumble project has a [guide](https://github.com/mumble-voip/mumble/blob/master/docs/dev/build-instructions/build_linux.md) for it.

**Comment by didierma34** [comment]
Hi,

I've found a package for Ubuntu and all works find but i've some problem to create special channel "fgcom-mumble".
I've read again server install but I don't understand how to create this special channel.
Couls you help me to create it ?
Thank's :) :)
Didier
 
 PS : prehaps after, i'll ask you some help to use GUI-RADIO... ;) ;) ;) but before I must do some tests with the special channel :) :)

**Reply by hbeni** [reply]
That is an mumble issue, not an fgcom one.
You need to be Admin on the mumble server, and there you just create a new, ordinary channel.
The only "special" thing about it is it's name, which needs to _start_ with the term `fgcom-mumble`.

Make sure, you have a mumble-server >=1.4.0. The ubuntu packages I know of, are still 1.3.4 (and thats not enough to support fgcom-mumble).

**Reply by didierma34** [reply]
:) Of course my version number is 1.3.4

So I've build with cmake and make... It's all right

But now, how to install my build ? ;) ;)

**Reply by hbeni** [reply]
You should end up with binaries in the `build`folder. You now just configure the server with the server `ini`file (tough defaults shuld  be sufficient for testing) and then just call up the server binary, for example from a shell: `mumble/build$ ./mumble-server`.
At startup the server prints the password of the user `SuperUser`into the console and logfile. You need to connect as `SuperUser` in order to add channels to the server.

You can also test against the "official" testing server at `fgcom.hallinger.org`.

**Comment by didierma34** [comment]
Ok...
How to know if mumble server is 1.4.0 ?

**Reply by hbeni** [reply]
Usually your package manager application (where you search mumble and push the "install"-button) tells you the version of the package.

**Comment by didierma34** [comment]
Hi,

All is all right now...
Plugin works well...
I've tried with GUIRadio like an ATC, but when I want to connect the GUIRadio with SimConnect, nothing works.
I've followed explanations in MD for GUIRadio, but I didn't what's wrong...
I think, there is no connection to SimConnect (I use MSFS 2020).

What can I try to test or to debug ?

Thank's :)

**Reply by hbeni** [reply]
When you activate the SimConnect bridge in the RadioGUI, that should make MSFS2020 transmit its values to the GUI.
When I wrote that piece, I did check against an actual MSFS2020 installation and it worked, the only thing not mapped was the PTT button (you need to use mumbles PTT feature).

But why do you want to d this? When you are ATCing, you are not supposed to sit in a plane, but to have separate radio equipment and possibly an dedicated ATC client.
The RadioGUI could be used standalone, just start it, set it up and hit the "connect" button in the lower left corner to make it send data to the mumble plugin.

Are we talking about the same RadioGUI java application that is shipped wit this project here?

**Reply by hbeni** [reply]
For reference, did you read https://github.com/hbeni/fgcom-mumble/blob/master/client/radioGUI/Readme.RadioGUI.md ?

**Reply by hbeni** [reply]
To be an ATC for MSFS2020, I think you should just start the radioGUI and mumble, and connect them. Thats all you should need.
Or use a dedicated ATC client (like ATCPie) instead of the radioGUI to also get a radar screen, etc.
(I'm not into MSFS, so I have no idea how ATC is normally/optimally done in that world)

**Reply by didierma34** [reply]
> When you activate the SimConnect bridge in the RadioGUI, that should make MSFS2020 transmit its values to the GUI. When I wrote that piece, I did check against an actual MSFS2020 installation and it worked, the only thing not mapped was the PTT button (you need to use mumbles PTT feature).
> 
Ok, I understand, but nothing works between MSFS and RadioGUI...

> But why do you want to d this? When you are ATCing, you are not supposed to sit in a plane, but to have separate radio equipment and possibly an dedicated ATC client. The RadioGUI could be used standalone, just start it, set it up and hit the "connect" button in the lower left corner to make it send data to the mumble plugin.
> 

In fact I believed that it was necessary to use RadioGUI also for the pilots. Indeed, my goal is to use the RadioGUI for an ATC only but then, what should be used for the pilots? Is Mumble with the FGCOM plugin enough for these drivers? The goal is to retrieve the frequencies displayed in the aircraft to speak to ATC. Or what do pilots need to have to speak with ATC using aircraft frequencies? I would allow to make a comparison with vPilot of VATSIM which recovers the frequencies of the plane to then speak to the ATC.
Thank you again for your help because it's really simple from you and congratulations again for your work. I just need to understand your principle of operation :) :) :)


> Are we talking about the same RadioGUI java application that is shipped wit this project here?

Yes, yes, We talk about your RAdioGUI project... :) :)

**Reply by didierma34** [reply]
> To be an ATC for MSFS2020, I think you should just start the radioGUI and mumble, and connect them. Thats all you should need. Or use a dedicated ATC client (like ATCPie) instead of the radioGUI to also get a radar screen, etc. (I'm not into MSFS, so I have no idea how ATC is normally/optimally done in that world)

Yes, of course, that's right.
But as I said in my previous answer, I now miss knowing that pilots must have in MSFS to be able to contact the ATC by choosing the frequencies in the plane. :) :) :)

**Reply by didierma34** [reply]
> For reference, did you read https://github.com/hbeni/fgcom-mumble/blob/master/client/radioGUI/Readme.RadioGUI.md ?

Yes, yes, I have read correctly and I have followed the instructions. :)

**Reply by hbeni** [reply]
> But as I said in my previous answer, I now miss knowing that pilots must have in MSFS to be able to contact the ATC by choosing the frequencies in the plane. :) :) :)

A MSFS Pilot is supposed to:
- run the mumble plugin
- run the RadioGUI in SimConnect-bridge mode
- run MSFS

I tested it with the default cessna 172, maybe that's also plane dependant?

Bad news:
I cannot test anymore because I know noone with an installation.
Also, I really have no big clue about SimConnect, I was glad it worked when it did - so a absolutely good idea would be to have a dedicated bridge application between MSFS and fgcom-mumble (the UDP data protocol is openly documented and quite easy; RadioGUI serves some sort of reference implementation). I don't see this as a main concern for *this* project however, and RadioGUI started out as kind of reference debug client tool for me :)
Maybe you have some grip at Java and SimConnect and can debug yourself?

**Reply by hbeni** [reply]
(Maybe this justifies a new separate discussion thread)

**Reply by didierma34** [reply]
ok, what I can give you is the source of a small program made with Visual Studio which allows you to retrieve information from the simulator via SimConnect.

I could actually try to debug Java but I'm not very knowledgeable about this language.

In fact, I don't really know the principle you use for radio communications, but I would be interested in trying to advance this ATC program for FGCOM, both for a pilot and for the controller.

**Reply by didierma34** [reply]
> (Maybe this justifies a new separate discussion thread)

Why not ? :) :) :)

**Reply by hbeni** [reply]
> ok, what I can give you is the source of a small program made with Visual Studio which allows you to retrieve information from the simulator via SimConnect.

This does not help me - I work on linux and have no visual studio.

> I could actually try to debug Java but I'm not very knowledgeable about this language.
> In fact, I don't really know the principle you use for radio communications, but I would be interested in trying to advance this ATC program for FGCOM, both for a pilot and for the controller.

The project here is perfectly suited and capable for that scenario already.
What exactly do you mean with your question "I don't really know the principle you use for radio communications"?

The overall architecture is documented here: https://github.com/hbeni/fgcom-mumble/blob/master/Readme.architecture.md
Basicly there is the fgcom-mumble mumble plugin. It adjusts the incoming audio streams based on internal data (add noise, cut out audio out of range). The data is established from two sources: A) other plugins send their metadata (radio stack, location etc) and B) UDP data send from a connected client (like flightgear, or the radioGUI) which provides _your own_ metadata (radio stack, location, etc).
Any program that sends UDP packets in the [proper format](https://github.com/hbeni/fgcom-mumble/blob/master/client/plugin.spec.md#plugin-input-data) can supply that data to the plugin. You could very well make a new program in visual studio that bridges MSFS to the plugin this way.

**Reply by didierma34** [reply]
Which Java IDE do you use ?

**Reply by hbeni** [reply]
See https://github.com/hbeni/fgcom-mumble/blob/master/client/radioGUI/Readme.RadioGUI.md#compiling
> The client is written using netbeans

Netbeans 12.0 I used.

**Reply by hbeni** [reply]
SimConnect code is here: https://github.com/hbeni/fgcom-mumble/blob/master/client/radioGUI/src/main/java/hbeni/fgcom_mumble/SimConnectBridge.java

**Reply by didierma34** [reply]
Ok... I take a look :) :)

**Reply by didierma34** [reply]
I've installed NetBeans, java, java SDK, etc... allright

I've cloned GitHub project of RadioGUI.

When I debug in NetBean, I've an error : 

-----------------< fgcom-mumble:FGCom-mumble-radioGUI >-----------------
Building FGCom-mumble-radioGUI 1.0.0
--------------------------------[ jar ]---------------------------------
The POM for flightsim:jsimconnect:jar:0.8.0 is missing, no dependency information available
The POM for org.openstreetmap.jmapviewer:jmapviewer:jar:2.14 is missing, no dependency information available
------------------------------------------------------------------------
BUILD FAILURE
------------------------------------------------------------------------
Total time:  0.144 s
Finished at: 2022-05-17T17:11:04+02:00
------------------------------------------------------------------------
Failed to execute goal on project FGCom-mumble-radioGUI: Could not resolve dependencies for project fgcom-mumble:FGCom-mumble-radioGUI:jar:1.0.0: The following artifacts could not be resolved: flightsim:jsimconnect:jar:0.8.0, org.openstreetmap.jmapviewer:jmapviewer:jar:2.14: flightsim:jsimconnect:jar:0.8.0 was not found in https://repo.maven.apache.org/maven2 during a previous attempt. This failure was cached in the local repository and resolution is not reattempted until the update interval of central has elapsed or updates are forced -> [Help 1]

To see the full stack trace of the errors, re-run Maven with the -e switch.
Re-run Maven using the -X switch to enable full debug logging.

For more information about the errors and possible solutions, please read the following articles:
[Help 1] http://cwiki.apache.org/confluence/display/MAVEN/DependencyResolutionException
-----------------------------------------------------------------------------------------------------


Do you know why ?

**Reply by hbeni** [reply]
Yes, of course.
Jsimconnect is referenced as git submodule, so you need to fetch the submodules after you checked out the git repo.
Jsimconnect could also be found at: https://github.com/mharj/jsimconnect

**Reply by didierma34** [reply]
Ok...

How to fetch a submodule in NetBeans ?
Thank's

**Reply by didierma34** [reply]
I've gone to jsimconnect but there is no "pom" in project. 
So I don't how to fetch submodule with RadioGUI.... Sorry :( :(

**Reply by hbeni** [reply]
How did you got the source?
Did you made a `git clone`? then the shell command would be `git submodule update --init --recursive`. jsimconnect is then fetched and build togheter with the main build process (`make build-radioGUI` which will install the project as local repo).
Otherwise you need to fetch and setup everything manually.

[edit] you checkout the sources on the command line OR a git client OR manually download the source tree from github

The build is made to work on linux, on windows I can't help much unfortunately.

**Reply by didierma34** [reply]
> The build is made to work on linux, on windows I can't help much unfortunately.

I'm on Windows ! :) :) :)

Never mind... :( :( :(
I won't be able to try debugging in Java... snif snif :) :) ;) 

I'll look at your code in NetBeans and I'll try with Windev (PCSOFT) to connect to fgcom-plugin in UDP (port 16661) with a right message. :) :)
I've made a connector to SimConnect and I use it. I know frequencies, so i'll try this.
In the end, it will do the same thing as RadioGUI for the ATC and for the pilot.

I'll say you if it's work :) :) :)

**Reply by didierma34** [reply]
So could tell me informations about UDP Message ?

CALLSIGN is a string for Callsign ? (it's not difficult :) :) :)  )
LAT and LON are Latitude and Longitude, but are they in decimals ?
HGT is Altitude above ground Level, but in feet or meters ?
COM1_PTT (Push To Talk) is 1 or 0 ?
COM1_FRQ is frequency in Khz or in Htz ? ( exemple : 121.250 is 121500 ?)
COM1_VOL is an int but wich values ? I think it's 0 to ???
COM1_PWR is the Power of Transmitter in Watt ? How is it for a Plane ?
COM1_SQC is for the Squelch ? but which values ?
COM1_CWKHZ is Channel Wide ? 8.33 or which other values ?
COM1_PBT is for Power Button ? is it 0 or 1 ?   0 = Off and 1 = On ?
AUDIO_FX_RADIO is 1 for On and 0 for Off ?
AUDIO_HEAR_ALL is 1 for On and 0 for Off (Off = hear only on the same frequency) ?

and DEBUG_SIGQLY, I think i'ts for debugging signal, but how to use it ?

Thank's :) :) :)

**Reply by hbeni** [reply]
> I'll say you if it's work :) :) :)

would be very cool if you could opensource this, so others can benefit from your work :)

> I'll look at your code in NetBeans and I'll try with Windev (PCSOFT) to connect to fgcom-plugin in UDP (port 16661) with a right message. :) :)

In the end it is straightforward. You just need some piece of software that continuously read out MSFS and craft UPD packages in a "special" string format, that will then be sent to the plugins UDP port (this is what the "bridge mode" was supposed to deliver and what worked when i tested it).
From github actions, you should be able to download a nightly build in debug mode. This prints out the received and processed UDP messages of the plugin, alongside the plugins internal state every second. This might come in handy to see if your messages are parsed correctly.
But the key information is also displayed in your mumble's identity comment.

The needed [UDP format is extensively specified in the doc's](https://github.com/hbeni/fgcom-mumble/blob/master/client/plugin.spec.md#plugin-input-data), i think you already found it.
To get real world examples, fire up the RadioGUI and look at the string next to the "connect" button, there the gui prints out what is actually being sent to the plugin: 
![grafik](https://user-images.githubusercontent.com/13608602/169069136-96944259-244b-4bbb-a196-5337db6d856d.png)

`CALLSIGN=ZZZZ,LAT=48.3440238,LON=11.765083,HGT=6.6,COM1_PTT=0,COM1_FRQ=121.000,COM1_VOL=1.0,COM1_PWR=10.0,COM1_SQC=0.1,COM1_CWKHZ=8.33,COM1_PBT=1,COM2_PTT=0,COM2_FRQ=122.540,COM2_VOL=1.0,COM2_PWR=10.0,COM2_SQC=0.1,COM2_CWKHZ=8.33,COM2_PBT=1,DEBUG_SIGQLY=-1.0,AUDIO_FX_RADIO=1,AUDIO_HEAR_ALL=0 `

**Reply by hbeni** [reply]
> So could tell me informations about UDP Message ?

RTFM :) (see post above for details)

**Reply by didierma34** [reply]
> The needed [UDP format is extensively specified in the doc's](https://github.com/hbeni/fgcom-mumble/blob/master/client/plugin.spec.md#plugin-input-data), i think you already found it.

Sorry, I didn't see that... All right. 
Thank's !!!!!!!!!!!

Yes, of course it will be open source, I could tell you code but it's very special and in French cause I'm a old Frenchy :) :) :) :) I'm 59 years old.

**Comment by hbeni** [comment]
> and DEBUG_SIGQLY, I think i'ts for debugging signal, but how to use it ?

Its meant for debugging radio models and should not be a concern for your case.
Anyway, specifying `DEBUG_SIGQLY=<FLOAT>` (ex. `DEBUG_SIGQLY=0.72`) will get ignored in a normal plugin build.
A debug build of the plugin will force any transmission to be received at the given signal quality, regardless the calculated quality of the radio model in question. You can use this if you want to fine tune the noise/cancellation audio effects for a given radio model and signal quality.

**Reply by didierma34** [reply]
Ok

**Comment by didierma34** [comment]
Hi,
I have a little problem with the message that is transmitted to the plugin.  
I saw this code: 

byte[] buffer = msg.getBytes();  
DatagramPacket request = new DatagramPacket(buffer, buffer.length, address, radioGUI.Options.udpPort); 

Can you explain this message to me ?  
Are you converting the string into a byte array ?  
Can you tell me what this DatagramPacket is?  
All this is useful for me to know how and in what form the message (msg) is transmitted.

Thank's

**Reply by hbeni** [reply]
Yes, just send the string as bytes in plaintext over the wire.

**Reply by hbeni** [reply]
> Can you tell me what this DatagramPacket is?

Its a standard java lang class: https://docs.oracle.com/javase/7/docs/api/java/net/DatagramPacket.html
It models UDP datagram packets, the standard UDP network format that takes care of the UDP headers etc. Basicly you use it to package up your data into a UDP network message which then can be `send()`to the network card (via the class `DatagramSocket`).

**Reply by didierma34** [reply]
Ok... thank's

**Reply by didierma34** [reply]
Hi,

I just made a small piece of code to try to send a message via a UDP socket, but the plugin does not respond to me and nothing is displayed on Mumble (A message on the left on Mumble is displayed when RadioGUI connects and when he sends his first message).

So I have a question:
I can't create a socket to address 127.0.0.1 and I have to use my PC's address. It is a problem related to the language which does not allow the creation of a UDP Socket with the address 127.0.0.1
Is it absolutely necessary to send the packet to 127.0.0.1 or is it possible by addressing the packet to the IP of the PC ?

**Reply by hbeni** [reply]
> I can't create a socket to address 127.0.0.1 and I have to use my PC's address. It is a problem related to the language which does not allow the creation of a UDP Socket with the address 127.0.0.1

I have no Idea, but that sounds weird to me and an issue with your computer/OS setup. Java definitely allows sending to arbitary IP's.

> Is it absolutely necessary to send the packet to 127.0.0.1 or is it possible by addressing the packet to the IP of the PC ?

As long as the listening socket is open, it accepts messages. By default it listens on `127.0.0.1` which should ALWAYS be reachable- its the local loopback interface address. Are you employing a local firewall application?

You ca instruct the plugin to listen on any adress by supplying a config file, details [here](https://github.com/hbeni/fgcom-mumble/blob/master/README.md#plugin-configuration), and:
See: https://github.com/hbeni/fgcom-mumble/blob/df16cc6313c221694f38516733dd75a9d8ac2b41/client/mumble-plugin/fgcom-mumble.ini#L63-L74

**Comment by didierma34** [comment]
Hi (Sorry I don't know your first name, mopi is Didier)

I have good news !!! :) :) :)

I just made a piece of code and I am connected with MSFS + My Application + Mumble with your Plugin and everything works !!

I can change the frequency of the plane so I can interact with Mumble for connections.

I continue my tests with more features and I'll let you know.

:) :)

**Comment by didierma34** [comment]
Hello,

I've now a little problem with Mumble :
When I launch Mumble with FGcom, Mumble crash every time during launching and I must send (or not) a report to continue.
When I launch Mumble WITHOUT FGcom, all is all right and Mumble works fine (no crash).

Do you know what can I do ?
Thank's

**Reply by hbeni** [reply]
No, but that’s a known issue in the mumble project.
It also happens sometimes without fgcom-mumble plugin.

however, if you can reproduce the crash, use debug builds of mumble and fgcom-mumble and produce a stack trace.

**Reply by didierma34** [reply]
I just find why :
When Mumble stops, it creates a file "mumble.dmp".
If you delete this file before launching Mumble, there is no crash :) :) :)
